### PR TITLE
Grading student answers in notebook

### PR DESCRIPTION
**Pull request by wilkosch**
Allows for in notebook testing of whether the students answers are correct. Adds grader class which also contains some utility methods to for running and getting locals from notebooks, necessary for gradescope.

For example by including the following code at the end of the vibrating_pendulum.ipynb.
`from underactuated.exercises.pend.vibrating_pendulum.test_vibrating_pendulum import TestVibratingPendulum
from underactuated.exercises.grader import Grader
Grader.grade_output([TestVibratingPendulum], [locals()], 'results.json')
Grader.print_test_results('results.json')`
Adding this code snippet to all relevant notebooks will be done in a separate PR, once #291 and #304 are merged.

This PR already contains all necessary tests.

---
This change is [Reviewable](https://reviewable.io/reviews/russtedrake/underactuated/305)

### CODE REVIEW

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import numpy as np
+
+
+class TestCartPoleInBal(unittest.TestCase):
+
+    def __init__(self, test_name, notebook_locals):
+        super().__init__(test_name)
+        self.notebook_locals = notebook_locals
+
+    @weight(8)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: https://github.com/RussTedrake/underactuated/pull/305/files?file-filters%5B%5D=.py

Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 2 points

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    def test_system_recovers_from_states(self):
+        '''"Test for which the system is able to recover"'''
+        # note: all prints here go to the output item in the json file
+        system_recovers_from_states = self.notebook_locals['system_recovers_from_states']
+        arr = np.sort(np.asarray(system_recovers_from_states))
+        a = hash(tuple(arr))
+        self.assertEqual(a, 6260030330839110161,
+                         "Incorrect states from which the system is able to recover.")
+
+    @weight(4)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 2 points

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    def test_get_A_lin(self):
+        '''"Test linearization matrix A"'''
+        # note: all prints here go to the output item in the json file
+        get_A_lin = self.notebook_locals['get_A_lin']
+        A_lin = get_A_lin()
+        a = hash(tuple(np.ndarray.flatten(A_lin)))
+        self.assertEqual(a, -4269322539335713771,
+                         "Incorrect linearization matrix A.")
+
+    @weight(4)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 2 points

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    def test_get_B_lin(self):
+        '''"Test linearization matrix B"'''
+        # note: all prints here go to the output item in the json file
+        get_B_lin = self.notebook_locals['get_B_lin']
+        B_lin = get_B_lin()
+        a = hash(tuple(np.ndarray.flatten(B_lin)))
+        self.assertEqual(a, 4686582722430018711,
+                         "Incorrect linearization matrix B.")
+
+    @weight(4)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 2 points

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        for i in range(len(u_list)):
+            self.assertLess(errors[i], ub[i],
+                            'Some linearization errors are too high')
+
+        lb = [0.00001, 1.0, 1.0, 23, 60, 0.00]
+        for i in range(len(u_list)):
+            self.assertLessEqual(
+                lb[i], errors[i], 'Some linearization errors are too low')
+
+    @weight(8)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 4 points

File: underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        for i in range(n_rands):
+            x = np.random.rand(4)
+            u = np.random.rand(1)
+            f_eval.append(f(x, u))
+        # print(repr(np.stack(f_eval)))
+        # print(np.linalg.norm(f_target - np.stack(f_eval)))
+        self.assertLessEqual(np.linalg.norm(f_target - np.stack(f_eval)), 1e-5,
+                             'The state space dynamics f(x,u) are not correct.')
+
+    @weight(2)
```
Comment (TobiaMarcucci) [inline underactuated/exercises/acrobot/cartpole_in_bal/test_cartpole_in_bal.py]: 1 point

**Comment by wilkosch** [comment]
> 
> 
> can you please add these files into their corresponding BUILD files, so that they get vetted by CI. (including linting)
> 

Done!

> > Reviewed 3 of 6 files at r1, 4 of 4 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/russtedrake/underactuated/305#-:-M0lZWyf6H50MaentiOQ:bwhz6xo)_ status: all files reviewed, 6 unresolved discussions (waiting on @wilkosch)
> 
> _[underactuated/exercises/acrobot/cartpole_balancing/test_cartpole_balancing.py, line 129 at r2](https://reviewable.io/reviews/russtedrake/underactuated/305#-M0lYM7V4sZkwdXKl_vy:-M0lYM7V4sZkwdXKl_vz:b-8vxmag) ([raw file](https://github.com/russtedrake/underactuated/blob/e223fd8dfd0ab8749bd563dfd823e5e0b9eb2aeb/underactuated/exercises/acrobot/cartpole_balancing/test_cartpole_balancing.py#L129)):_
> 
> > ```python
> >             f_eval.append(f(x, u))
> >         # print(repr(np.stack(f_eval)))
> >         # print(np.linalg.norm(f_target - np.stack(f_eval)))
> > ```
> 
> fwiw, can probably zap these?

Done!

**Comment by wilkosch** [comment]
> 
> 
> > Reviewed 9 of 9 files at r3.
> > _[Reviewable](https://reviewable.io/reviews/russtedrake/underactuated/305)_ status: all files reviewed, 7 unresolved discussions (waiting on @wilkosch)
> 
> _[underactuated/exercises/acrobot/cartpole_balancing/BUILD.bazel, line 12 at r3](https://reviewable.io/reviews/russtedrake/underactuated/305#-M1IqGEeE0yHVDOQg2JO:-M1IqGEeE0yHVDOQg2JP:b-rqx4qo) ([raw file](https://github.com/russtedrake/underactuated/blob/b3f15877c4a1a85c89cc6ef9341fe4f4bb9bc418/underactuated/exercises/acrobot/cartpole_balancing/BUILD.bazel#L12)):_
> 
> > ```python
> >     name = "test_cartpole_balancing",
> >     srcs = ["test_cartpole_balancing.py"],
> >     size = "enormous",
> > ```
> 
> that's a bad sign. if we have too many like this, it will really slow down the testing (including CI). is it justified? in many of my examples, I run less of them for testing (by checking for ipython, etc).

Good catch. It is not necessary and was a remnant of some playing around with bazel. Sorry!
I removed that line

**Pull request merged**